File: ios/Sources/AgentlyAppFoundation/Support/VisibleContentSanitizer.swift
import Foundation

private let hiddenRouterPayloadKeys: Set<String> = [
    "appendToolBundles",
    "suggestedProfileId",
    "templateId",
    "classification",
    "prompting",
    "directAction",
    "scope",
    "clarificationNeeded",
    "clarificationQuestion"
]

func sanitizeVisibleAssistantText(_ raw: String?) -> String? {
    let trimmed = raw?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    guard !trimmed.isEmpty else { return nil }
    let stripped = stripHiddenRouterPayload(from: trimmed)
    return stripped.isEmpty ? nil : stripped
}

private func stripHiddenRouterPayload(from text: String) -> String {
    if isHiddenRouterPayloadJSON(text) {
        return ""
    }

    guard let start = text.firstIndex(of: "{"),
          let end = text.lastIndex(of: "}"),
          start < end else {
        return text
    }

    let candidate = String(text[start...end]).trimmingCharacters(in: .whitespacesAndNewlines)
    guard isHiddenRouterPayloadJSON(candidate) else {
        return text
    }

    let prefix = String(text[..<start]).trimmingCharacters(in: .whitespacesAndNewlines)
    return prefix
}

private func isHiddenRouterPayloadJSON(_ text: String) -> Bool {
    guard let data = text.data(using: .utf8),
          let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
        return false
    }
    return !hiddenRouterPayloadKeys.isDisjoint(with: object.keys)
}

File: android/app/src/main/java/com/viant/agently/android/VisibleContentSanitizer.kt
package com.viant.agently.android

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject

private val hiddenRouterPayloadKeys = setOf(
    "appendToolBundles",
    "suggestedProfileId",
    "templateId",
    "classification",
    "prompting",
    "directAction",
    "scope",
    "clarificationNeeded",
    "clarificationQuestion"
)

private val visibleContentJson = Json { ignoreUnknownKeys = true }

internal fun sanitizeVisibleAssistantText(raw: String?): String? {
    val trimmed = raw?.trim().orEmpty()
    if (trimmed.isEmpty()) return null
    val stripped = stripHiddenRouterPayload(trimmed)
    return stripped.takeIf { it.isNotEmpty() }
}

private fun stripHiddenRouterPayload(text: String): String {
    if (isHiddenRouterPayloadJson(text)) return ""
    val start = text.indexOf('{')
    val end = text.lastIndexOf('}')
    if (start < 0 || end <= start) return text
    val candidate = text.substring(start, end + 1).trim()
    if (!isHiddenRouterPayloadJson(candidate)) return text
    return text.substring(0, start).trim()
}

private fun isHiddenRouterPayloadJson(text: String): Boolean {
    val payloadObject = runCatching {
        visibleContentJson.parseToJsonElement(text).jsonObject
    }.getOrNull() ?: return false
    return payloadObject.keys.any(hiddenRouterPayloadKeys::contains)
}
